File: core/src/main/java/com/mcreater/amclcore/command/StartProcessTask.java
package com.mcreater.amclcore.command;

import com.mcreater.amclcore.concurrent.ConcurrentExecutors;
import com.mcreater.amclcore.concurrent.ExtendForkJoinPool;
import com.mcreater.amclcore.concurrent.TaskState;
import com.mcreater.amclcore.concurrent.task.AbstractTask;
import com.mcreater.amclcore.concurrent.task.model.RunnableAction;
import com.mcreater.amclcore.i18n.Text;
import lombok.Getter;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.List;
import java.util.Optional;
import java.util.Vector;
import java.util.function.BiConsumer;
import java.util.stream.Collectors;

import static com.mcreater.amclcore.concurrent.ConcurrentExecutors.excHandler;
import static com.mcreater.amclcore.i18n.I18NManager.fixed;
import static com.mcreater.amclcore.i18n.I18NManager.translatable;

public class StartProcessTask extends AbstractTask<Integer> {
    public static StartProcessTask create(List<CommandArg> args) {
        return new StartProcessTask(args);
    }

    private StartProcessTask(List<CommandArg> args) {
        this.args = args;
    }

    private final List<CommandArg> args;
    private Path startPath;
    private Process process;
    @Getter
    private final List<OutputParser.OutputLine> logLines = new Vector<>();
    private final ExtendForkJoinPool pool = new ExtendForkJoinPool(
            2,
            ConcurrentExecutors.ForkJoinWorkerThreadFactoryImpl.INSTANCE,
            excHandler,
            true
    );

    public StartProcessTask setStartPath(Path path) {
        this.startPath = path;
        return this;
    }

    protected Integer call() throws Exception {
        setState(
                TaskState.<Integer>builder()
                        .totalStage(2)
                        .currentStage(0)
                        .build()
        );
        process = new ProcessBuilder().command(
                        args.stream()
                                .map(CommandArg::toString)
                                .collect(Collectors.toList())
                )
                .directory(startPath.toFile())
                .start();
        Runtime.getRuntime().addShutdownHook(new Thread(this.process::destroy));

        BufferedReader in = new BufferedReader(new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8));
        BufferedReader err = new BufferedReader(new InputStreamReader(process.getErrorStream(), StandardCharsets.UTF_8));

        BiConsumer<String, PrintStream> consumer = (s, ps) -> {
            Optional.of(OutputParser.OutputLine.create(s, ps)).map(a -> {
                logLines.add(a);
                return a;
            }).get().printAnsi();
            setState(
                    TaskState.<Integer>builder()
                            .totalStage(2)
                            .currentStage(1)
                            .message(fixed(s))
                            .build()
            );
        };

        RunnableAction.of(() -> {
            do {
                try {
                    Optional.ofNullable(in.readLine()).ifPresent(a -> {
                        consumer.accept(a, System.out);
                    });
                } catch (IOException ignored) {
                }
            }
            while (process.isAlive());
        }).submitTo(pool);

        RunnableAction.of(() -> {
            do {
                try {
                    Optional.ofNullable(err.readLine()).ifPresent(a -> {
                        consumer.accept(a, System.err);
                    });
                } catch (IOException ignored) {
                }
            }
            while (process.isAlive());
        }).submitTo(pool);

        do {
        }
        while (process.isAlive());

        setState(
                TaskState.<Integer>builder()
                        .totalStage(2)
                        .currentStage(2)
                        .message(translatable("core.process.exit", process.exitValue()))
                        .build()
        );

        return process.exitValue();
    }

    protected Text getTaskName() {
        return translatable("core.task.process.name");
    }
}

File: core/src/main/java/com/mcreater/amclcore/command/StartProcessTask.kt
package com.mcreater.amclcore.command

import com.mcreater.amclcore.command.OutputParser.OutputLine
import com.mcreater.amclcore.command.OutputParser.OutputLine.Companion.create
import com.mcreater.amclcore.concurrent.ConcurrentExecutors
import com.mcreater.amclcore.concurrent.ExtendForkJoinPool
import com.mcreater.amclcore.concurrent.TaskState
import com.mcreater.amclcore.concurrent.task.AbstractTask
import com.mcreater.amclcore.concurrent.task.model.RunnableAction
import com.mcreater.amclcore.i18n.I18NManager.Companion.fixed
import com.mcreater.amclcore.i18n.I18NManager.Companion.translatable
import com.mcreater.amclcore.i18n.Text
import lombok.Getter
import java.io.*
import java.nio.charset.StandardCharsets
import java.nio.file.Path
import java.util.*
import java.util.function.BiConsumer
import java.util.stream.Collectors


class StartProcessTask private constructor(private val args: List<CommandArg>) :
    AbstractTask<Int?>() {
    private var startPath: Path? = null
    private var process: Process? = null

    @Getter
    private val logLines: MutableList<OutputLine> = Vector()
    private val pool = ExtendForkJoinPool(
        2,
        ConcurrentExecutors.ForkJoinWorkerThreadFactoryImpl.INSTANCE,
        ConcurrentExecutors.excHandler,
        true
    )

    fun setStartPath(path: Path?): StartProcessTask {
        startPath = path
        return this
    }

    @Throws(Exception::class)
    override fun call(): Int {
        setState(
            TaskState(
                totalStage = 2,
                currentStage = 0
            )
        )
        process = ProcessBuilder().command(
            args.stream()
                .map { obj: CommandArg -> obj.toString() }
                .collect(Collectors.toList())
        )
            .directory(startPath!!.toFile())
            .start()
        Runtime.getRuntime().addShutdownHook(Thread { process?.destroy() })
        val input = BufferedReader(
            InputStreamReader(
                process?.inputStream ?: ByteArrayInputStream(ByteArray(1)),
                StandardCharsets.UTF_8
            )
        )
        val err = BufferedReader(
            InputStreamReader(
                process?.errorStream ?: ByteArrayInputStream(ByteArray(1)),
                StandardCharsets.UTF_8
            )
        )
        val consumer =
            BiConsumer { s: String?, ps: PrintStream? ->
                Optional.of(
                    create(
                        s,
                        ps!!
                    )
                ).map { a: OutputLine ->
                    logLines.add(a)
                    a
                }.get().printAnsi()
                setState(
                    TaskState(
                        totalStage = 2,
                        currentStage = 1,
                        message = fixed(s)
                    )
                )
            }
        RunnableAction.of {
            do {
                try {
                    Optional.ofNullable(input.readLine())
                        .ifPresent { a: String ->
                            consumer.accept(
                                a,
                                System.out
                            )
                        }
                } catch (ignored: IOException) {
                }
            } while (process?.isAlive == true)
        }.submitTo(pool)
        RunnableAction.of {
            do {
                try {
                    Optional.ofNullable(err.readLine())
                        .ifPresent { a: String ->
                            consumer.accept(
                                a,
                                System.err
                            )
                        }
                } catch (ignored: IOException) {
                }
            } while (process?.isAlive == true)
        }.submitTo(pool)
        do {
        } while (process?.isAlive == true)
        setState(
            TaskState(
                totalStage = 2,
                currentStage = 2,
                message = translatable("core.process.exit", process?.exitValue() ?: -1)
            )
        )
        return process?.exitValue() ?: -1
    }

    override fun getTaskName(): Text {
        return translatable("core.task.process.name")
    }

    companion object {
        @JvmStatic
        fun create(args: List<CommandArg>): StartProcessTask {
            return StartProcessTask(args)
        }
    }
}
